File: frontend/src/modules/notifications/interfaces/INotificationStrategy.ts
/**
 * Root contract for a notification delivery channel (frontend side).
 *
 * Responsibility (SRP): declare how a single channel delivers a notification.
 * Depends on: nothing — root abstraction.
 * Pattern: Strategy — each channel implements this interface.
 * SOLID: DIP · OCP · LSP · SRP
 *
 * OCP: SMSStrategy / PushStrategy = new class + registration in NotificationFactory.
 *      Existing strategies unchanged.
 *
 * Note: the FE strategy layer is thin — most delivery logic lives in the BE.
 *       FE strategies are used only for client-side channel simulation in tests.
 */

export interface INotificationStrategy {
  /** Check the channel can reach this recipient (e.g. browser permission for Push). */
  validate(recipientId: string): Promise<boolean>

  /** Deliver the notification payload via this channel. */
  send(recipientId: string, message: string, context: Record<string, unknown>): Promise<void>

  /** Record the delivery attempt result for debugging/observability. */
  log(status: 'sent' | 'failed' | 'skipped', details: string): void
}

File: backend/apps/notifications/interfaces/i_notification_strategy.py
"""
Root ABC for all notification delivery channels.

Responsibility (SRP): declare the contract for sending a notification via one channel.
    No routing logic, no preference checks — just the delivery contract.
Depends on: abc.ABC — nothing from the domain.
Pattern: Strategy — each channel is a concrete strategy behind this interface.
SOLID: DIP · OCP · LSP · SRP

OCP extension:
    SMSStrategy / PushStrategy = new class implementing INotificationStrategy
    + one entry in NotificationFactory.CHANNEL_MAP.
    Existing strategies (Email, InApp, WebSocket) are NEVER modified.

LSP:
    NotificationService always receives INotificationStrategy — any concrete
    strategy is substitutable in tests (inject a mock without touching the service).

Sprint coverage:
    S19 → this file (contract) + 3 concrete strategies
    S20 → NotificationService.dispatch() consumes this interface
    S27 → tests mock INotificationStrategy; strategies tested in isolation
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class INotificationStrategy(ABC):
    """Abstract contract for a single notification delivery channel."""

    @abstractmethod
    def validate(self, recipient) -> bool:
        """
        Check that the channel can deliver to this recipient.
        Examples:
            EmailStrategy: recipient.email is non-empty and verified.
            WebSocketStrategy: the user's WS group exists (always True — fire-and-forget).
            InAppStrategy: recipient is active.

        Args:
            recipient: User model instance.

        Returns:
            bool — False means skip this channel for this recipient (no exception).
        """
        ...

    @abstractmethod
    def send(self, recipient, message: str, context: dict) -> None:
        """
        Deliver the notification to the recipient via this channel.

        Args:
            recipient: User model instance (the notification destination).
            message:   Plain-text summary of the notification.
            context:   Channel-specific data dict.
                       EmailStrategy expects: {'tipo', 'subject', 'titulo', 'cuerpo', ...}
                       InAppStrategy expects: {'tipo', 'titulo', 'cuerpo', 'payload'}
                       WebSocketStrategy expects: {'notification_id', 'tipo', 'titulo', 'cuerpo'}

        Raises:
            NotificationDeliveryError — if the channel fails after internal retries.
                NotificationService catches this and continues with other channels.
        """
        ...

    @abstractmethod
    def log(self, status: str, details: str) -> None:
        """
        Record the delivery attempt result.

        Args:
            status:  'sent' | 'failed' | 'skipped'
            details: Human-readable explanation (channel, recipient email/id, error message).
        """
        ...
